README.md
# twemproxies (nutcrackers)

**twemproxies** ( **nutcrackers** ) is a multithread, fast and lightweight proxy for [memcached](http://www.memcached.org/) and [redis](http://redis.io/) protocol. It was built primarily to reduce the number of connections to the caching servers on the backend. This, together with protocol pipelining and sharding enables you to horizontally scale your distributed caching architecture.

### QQ群：129951966

## Build

To build twemproxies from source with _debug logs enabled_ and _assertions enabled_:

    $ git clone git@github.com:twitter/twemproxies.git
    $ cd twemproxies
    $ autoreconf -fvi
    $ ./configure --enable-debug=full
    $ make
    $ src/nutcrackers -h

A quick checklist:

+ Use newer version of gcc (older version of gcc has problems)
+ Use CFLAGS="-O1" ./configure && make
+ Use CFLAGS="-O3 -fno-strict-aliasing" ./configure && make
+ `autoreconf -fvi && ./configure` needs `automake` and `libtool` to be installed

## Note

Try the follow command to get the twemproxies status:
	
    printf "status\r\n" | nc manage_ip manage_port
    
## Performance

We test twemproxis performance on our online physical machine.

##### Machine Configuration:

OS: Centos6 Linux 2.6.32-504.23.4.el6.x86_64

Cpu Model Name: Intel(R) Xeon(R) CPU E5-2630 v2 @ 2.60GHz

Cpu Physical: 2

Cpu Cores: 6

Cpu Processors: 24

Network: 10000Mbps

Memory: 128G

##### Test Example:

Command: set

Key: one million random keys (used redis-benchmark -r option)

Key length: 16

Value length: 3

##### Test Results:

| twemproxies thread number | redis-benchmark number | client connections | qps(k/s) | cpu(%) |
| -------------------------:| ----------------------:| ------------------:| --------:| ------:|
| 1                         | 1                      | 8                  |       45 |     90 |
| 2                         | 2                      | 16                 |       84 |    187 |
| 3                         | 3                      | 24                 |      121 |    285 |
| 4                         | 4                      | 32                 |      150 |    385 |
| 5                         | 5                      | 40                 |      186 |    478 |
| 6                         | 6                      | 48                 |      210 |    583 |
| 7                         | 7                      | 56                 |      235 |    681 |
| 8                         | 8                      | 64                 |      264 |    780 |
| 9                         | 9                      | 72                 |      289 |    878 |
| 10                        | 10                     | 80                 |      313 |    976 |
| 11                        | 11                     | 88                 |      336 |   1074 |
| 12                        | 12                     | 96                 |      354 |   1175 |
| 13                        | 13                     | 104                |      366 |   1280 |
| 14                        | 14                     | 112                |      386 |   1380 |
| 15                        | 15                     | 120                |      405 |   1480 |
| 16                        | 16                     | 128                |      422 |   1580 |
| 17                        | 17                     | 136                |      442 |   1681 |
| 18                        | 18                     | 144                |      460 |   1779 |
| 19                        | 19                     | 152                |      474 |   1880 |
| 20                        | 20                     | 160                |      490 |   1973 |

## Features

+ Multithread.
+ Fast.
+ Lightweight.
+ Maintains persistent server connections.
+ Keeps connection count on the backend caching servers low.
+ Enables pipelining of requests and responses.
+ Supports proxying to multiple servers.
+ Supports multiple server pools simultaneously.
+ Shard data automatically across multiple servers.
+ Implements the complete [memcached ascii](notes/memcache.md) and [redis](notes/redis.md) protocol.
+ Easy configuration of server pools through a YAML file.
+ Supports multiple hashing modes including consistent hashing and distribution.
+ Can be configured to disable nodes on failures.
+ Observability via stats exposed on the manage port.
+ Works with Linux, *BSD, OS X and SmartOS (Solaris)

## Help

    Usage: nutcrackers [-?hVdDt] [-v verbosity level] [-o output file]
                       [-c conf file] [-s manage port] [-a manage addr]
                       [-i interval] [-p pid file] [-m mbuf size]
					   [-T worker threads number]

    Options:
      -h, --help             : this help
      -V, --version          : show version and exit
      -t, --test-conf        : test configuration for syntax errors and exit
      -d, --daemonize        : run as a daemon
      -D, --describe-stats   : print stats description and exit
      -v, --verbose=N        : set logging level (default: 5, min: 0, max: 11)
      -o, --output=S         : set logging file (default: stderr)
      -c, --conf-file=S      : set configuration file (default: conf/nutcrackers.yml)
      -s, --port=N           : set manage port (default: 22222)
      -a, --addr=S           : set manage ip (default: 0.0.0.0)
      -i, --interval=N       : set interval in msec (default: 30000 msec)
      -p, --pid-file=S       : set pid file (default: off)
      -m, --mbuf-size=N      : set size of mbuf chunk in bytes (default: 16384 bytes)
	  -T, --thread_num=N     : set the worker threads number (default: 8)

## Zero Copy

In twemproxies, all the memory for incoming requests and outgoing responses is allocated in mbuf. Mbuf enables zero-copy because the same buffer on which a request was received from the client is used for forwarding it to the server. Similarly the same mbuf on which a response was received from the server is used for forwarding it to the client.

Furthermore, memory for mbufs is managed using a reuse pool. This means that once mbuf is allocated, it is not deallocated, but just put back into the reuse pool. By default each mbuf chunk is set to 16K bytes in size. There is a trade-off between the mbuf size and number of concurrent connections twemproxies can support. A large mbuf size reduces the number of read syscalls made by twemproxies when reading requests or responses. However, with a large mbuf size, every active connection would use up 16K bytes of buffer which might be an issue when twemproxies is handling large number of concurrent connections from clients. When twemproxies is meant to handle a large number of concurrent client connections, you should set chunk size to a small value like 512 bytes using the -m or --mbuf-size=N argument.

## Configuration

twemproxies can be configured through a YAML file specified by the -c or --conf-file command-line argument on process start. The configuration file is used to specify the server pools and the servers within each pool that twemproxies manages. The configuration files parses and understands the following keys:

+ **listen**: The listening address and port (name:port or ip:port) or an absolute path to sock file (e.g. /var/run/nutcrackers.sock) for this server pool.
+ **hash**: The name of the hash function. Possible values are:
 + one_at_a_time
 + md5
 + crc16
 + crc32 (crc32 implementation compatible with [libmemcached](http://libmemcached.org/))
 + crc32a (correct crc32 implementation as per the spec)
 + fnv1_64
 + fnv1a_64
 + fnv1_32
 + fnv1a_32
 + hsieh
 + murmur
 + jenkins
+ **hash_tag**: A two character string that specifies the part of the key used for hashing. Eg "{}" or "$$". [Hash tag](notes/recommendation.md#hash-tags) enable mapping different keys to the same server as long as the part of the key within the tag is the same.
+ **distribution**: The key distribution mode. Possible values are:
 + ketama
 + modula
 + random
+ **timeout**: The timeout value in msec that we wait for to establish a connection to the server or receive a response from a server. By default, we wait indefinitely.
+ **backlog**: The TCP backlog argument. Defaults to 512.
+ **tcpkeepalive**: A boolean value that controls if tcp keepalive enabled. Defaults to false.
+ **tcpkeepidle**: The time value in msec that a connection is in idle, and then twemproxy check this connection whether dead or not. 
+ **tcpkeepcnt**: The number of tcpkeepalive attempt check if one idle connection dead times when the client always had no reply. 
+ **tcpkeepintvl**: The time value in msec that the interval between every tcpkeepalive check when the client always had no reply.
+ **preconnect**: A boolean value that controls if twemproxies should preconnect to all the servers in this pool on process start. Defaults to false.
+ **redis**: A boolean value that controls if a server pool speaks redis or memcached protocol. Defaults to false.
+ **redis_auth**: Authenticate to the Redis server on connect.
+ **redis_db**: The DB number to use on the pool servers. Defaults to 0. Note: twemproxies will always present itself to clients as DB 0.
+ **server_connections**: The maximum number of connections that can be opened to each server. By default, we open at most 1 server connection.
+ **auto_eject_hosts**: A boolean value that controls if server should be ejected temporarily when it fails consecutively server_failure_limit times. See [liveness recommendations](notes/recommendation.md#liveness) for information. Defaults to false.
+ **server_retry_timeout**: The timeout value in msec to wait for before retrying on a temporarily ejected server, when auto_eject_host is set to true. Defaults to 30000 msec.
+ **server_failure_limit**: The number of consecutive failures on a server that would lead to it being temporarily ejected when auto_eject_host is set to true. Defaults to 2.
+ **servers**: A list of server address, port and weight (name:port:weight or ip:port:weight) for this server pool.


For example, the configuration file in [conf/nutcrackers.yml](conf/nutcrackers.yml), also shown below, configures 5 server pools with names - _alpha_, _beta_, _gamma_, _delta_ and omega. Clients that intend to send requests to one of the 10 servers in pool delta connect to port 22124 on 127.0.0.1. Clients that intend to send request to one of 2 servers in pool omega connect to unix path /tmp/gamma. Requests sent to pool alpha and omega have no timeout and might require timeout functionality to be implemented on the client side. On the other hand, requests sent to pool beta, gamma and delta timeout after 400 msec, 400 msec and 100 msec respectively when no response is received from the server. Of the 5 server pools, only pools alpha, gamma and delta are configured to use server ejection and hence are resilient to server failures. All the 5 server pools use ketama consistent hashing for key distribution with the key hasher for pools alpha, beta, gamma and delta set to fnv1a_64 while that for pool omega set to hsieh. Also only pool beta uses [nodes names](notes/recommendation.md#node-names-for-consistent-hashing) for consistent hashing, while pool alpha, gamma, delta and omega use 'host:port:weight' for consistent hashing. Finally, only pool alpha and beta can speak the redis protocol, while pool gamma, deta and omega speak memcached protocol.

    alpha:
      listen: 127.0.0.1:22121
      hash: fnv1a_64
      distribution: ketama
      auto_eject_hosts: true
      redis: true
      server_retry_timeout: 2000
      server_failure_limit: 1
      servers:
       - 127.0.0.1:6379:1

    beta:
      listen: 127.0.0.1:22122
      hash: fnv1a_64
      hash_tag: "{}"
      distribution: ketama
      auto_eject_hosts: false
      timeout: 400
      redis: true
      servers:
       - 127.0.0.1:6380:1 server1
       - 127.0.0.1:6381:1 server2
       - 127.0.0.1:6382:1 server3
       - 127.0.0.1:6383:1 server4

    gamma:
      listen: 127.0.0.1:22123
      hash: fnv1a_64
      distribution: ketama
      timeout: 400
      backlog: 1024
      preconnect: true
      auto_eject_hosts: true
      server_retry_timeout: 2000
      server_failure_limit: 3
      servers:
       - 127.0.0.1:11212:1
       - 127.0.0.1:11213:1

    delta:
      listen: 127.0.0.1:22124
      hash: fnv1a_64
      distribution: ketama
      timeout: 100
      auto_eject_hosts: true
      server_retry_timeout: 2000
      server_failure_limit: 1
      servers:
       - 127.0.0.1:11214:1
       - 127.0.0.1:11215:1
       - 127.0.0.1:11216:1
       - 127.0.0.1:11217:1
       - 127.0.0.1:11218:1
       - 127.0.0.1:11219:1
       - 127.0.0.1:11220:1
       - 127.0.0.1:11221:1
       - 127.0.0.1:11222:1
       - 127.0.0.1:11223:1

    omega:
      listen: /tmp/gamma
      hash: hsieh
      distribution: ketama
      auto_eject_hosts: false
      servers:
       - 127.0.0.1:11214:100000
       - 127.0.0.1:11215:1

Finally, to make writing a syntactically correct configuration file easier, twemproxies provides a command-line argument -t or --test-conf that can be used to test the YAML configuration file for any syntax error.

## Observability

Observability in twemproxies is through logs and stats.


ChangeLog
 2016-8-3  deep  <diguo58@gmail.com>
    * twemproxies: version 1.2.1 release
      fix a bug about memory leak for mset command

 2016-7-15  deep  <diguo58@gmail.com>
    * twemproxies: version 1.2.0 release
      support 'delete key 0' for mc
      change default thread number
      add help, find_key and find_keys command for manage
      
 2016-5-10  deep  <diguo58@gmail.com>
    * twemproxies: version 1.1.0 release
      add show_twemproxies_stats tool
      support spin lock for stats
 
 2016-4-20  deep  <diguo58@gmail.com>
    * twemproxies: version 1.0.0 release
      multi-threads support
    
 2015-22-06  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.4.1 release
      redis_auth is only valid for a redis pool
      getaddrinfo returns non-zero +ve value on error
      fix-hang-when-command-only (charsyam)
      fix bug crash when get command without key and whitespace (charsyam)
      mark server as failed on protocol level transiet failures like -OOM, -LOADING, etc
      implemented support for parsing fine grained redis error response
      remove redundant conditional judgement in rbtree deletion (leo ma)
      fix bug mset has invalid pair (charsyam)
      fix bug mset has invalid pair (charsyam)
      temp fix a core on kqueue (idning)
      support "touch" command for memcached (panmiaocai)
      fix redis parse rsp bug (charsyam)
      SORT command can take multiple arguments. So it should be part of redis_argn() and not redis_arg0()
      remove incorrect assert because client could send data after sending a quit request which must be discarded
      allow file permissions to be set for UNIX domain listening socket (ori liveneh)
      return error if formatted is greater than mbuf size by using nc_vsnprintf() in msg_prepend_format()
      fix req_make_reply on msg_get, mark it as response (idning)
      redis database select upon connect (arne claus)
      redis_auth (charsyam)
      allow null key(empty key) (idning)
      fix core on invalid mset like "mset a a a" (idning)

2014-18-10  idning <idning@gmail.com>
    * twemproxy: version 0.4.0 release
      mget improve (idning)
      many new commands supported: LEX, PFADD, PFMERGE, SORT, PING, QUIT, SCAN... (mattrobenolt, areina, idning)
      handle max open file limit(allenlz)
      add notice-log and use ms time in log(idning)
      fix bug in string_compare (andyqzb)
      fix deadlock in sighandler (idning)

2013-20-12  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.3.0 release
      SRANDMEMBER support for the optional count argument (mkhq)
      Handle case where server responds while the request is still being sent (jdi-tagged)
      event ports (solaris/smartos) support
      add timestamp when the server was ejected
      support for set ex/px/nx/xx for redis 2.6.12 and up (ypocat)
      kqueue (bsd) support (ferenyx)
      fix parsing redis response to accept integer reply (charsyam)

2013-23-04  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.2.4 release
      redis keys must be less than mbuf_data_size() in length (fifsky)
      Adds support for DUMP/RESTORE commands in Redis (remotezygote)
      Use of the weight value in the modula distribution (mezzatto)
      Add support to unix socket connections to servers (mezzatto)
      only check for duplicate server name and not 'host:port:weight' when 'name' is configured
      crc16 hash support added (mezzatto)

2013-31-01  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.2.3 release
      RPOPLPUSH, SDIFF, SDIFFSTORE, SINTER, SINTERSTORE, SMOVE, SUNION, SUNIONSTORE, ZINTERSTORE, and ZUNIONSTORE support (dcartoon)
      EVAL and EVALSHA support (ferenyx)
      exit 1 if configuration file is invalid (cofyc)
      return non-zero exit status when nutcracker cannot start for some reason
      use server names in stats (charsyam)
      Fix failure to resolve long FQDN name resolve (conmame)
      add support for hash tags

2012-18-10  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.2.2 release
      fix the off-by-one error when calculating redis key length

2012-12-10  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.2.1 release
      don't use buf in conf_add_server
      allow an optional instance name for consistent hashing (charsyam)
      add --stats-addr=S option
      add stats-bind-any -a option (charsyam)

2012-12-03  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.2.0 release
      add -D or --describe-stats command-line argument to print stats description
      redis support in twemproxy
      setup pre/post splitcopy and pre/post coalesce handlers in msg struct
      memcache pre_splitcopy, post_splitcopy, pre_coalesce and post_coalesce handlers
      every fragment of a msg vector keeps track of the first/last fragment, number of fragments and fragment owner
      set up msg parser handler for memcache connections
      refactor parsing code and create header file nc_proto.h
      stats_listen should use st->addr as the listening address string
      delete stats tracking memcache requests and responses; stats module no longer tracks protocol related stats

2012-10-27  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.1.20 release
      on msg_repair, msg->pos should point to nbuf->pos and not nbuf->last
      refactor memcache parsing code into proto directory
      add redis option to configuration file
      fix macro definition strXcmp error for big endian
      fix log_hexdump and loga_hexdump

2012-07-31  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.1.19 release
      close server connection on a stray response (yashh, bmatheny)

2012-06-19  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.1.18 release
      command line option to set mbuf chunk size

2012-05-09  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.1.17 release

src/nc.c
/*
 * twemproxy - A fast and lightweight proxy for memcached protocol.
 * Copyright (C) 2011 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <getopt.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/utsname.h>

#include <nc_core.h>
#include <nc_conf.h>
#include <nc_signal.h>

#define NC_CONF_PATH        "conf/nutcrackers.yml"

#define NC_LOG_DEFAULT      LOG_NOTICE
#define NC_LOG_MIN          LOG_EMERG
#define NC_LOG_MAX          LOG_PVERB
#define NC_LOG_PATH         NULL

#define NC_PORT             22222
#define NC_ADDR             "0.0.0.0"
#define NC_INTERVAL         (30 * 1000) /* in msec */

#define NC_PID_FILE         NULL

#define NC_MBUF_SIZE        MBUF_SIZE
#define NC_MBUF_MIN_SIZE    MBUF_MIN_SIZE
#define NC_MBUF_MAX_SIZE    MBUF_MAX_SIZE

#define NC_THREAD_NUM_DEFAULT	(sysconf(_SC_NPROCESSORS_ONLN)>6?6:sysconf(_SC_NPROCESSORS_ONLN))

static int show_help;
static int show_version;
static int test_conf;
static int daemonize;
static int describe_stats;

static struct option long_options[] = {
    { "help",           no_argument,        NULL,   'h' },
    { "version",        no_argument,        NULL,   'V' },
    { "test-conf",      no_argument,        NULL,   't' },
    { "daemonize",      no_argument,        NULL,   'd' },
    { "describe-stats", no_argument,        NULL,   'D' },
    { "verbose",        required_argument,  NULL,   'v' },
    { "output",         required_argument,  NULL,   'o' },
    { "conf-file",      required_argument,  NULL,   'c' },
    { "port",           required_argument,  NULL,   's' },
    { "interval",       required_argument,  NULL,   'i' },
    { "addr",           required_argument,  NULL,   'a' },
    { "pid-file",       required_argument,  NULL,   'p' },
    { "mbuf-size",      required_argument,  NULL,   'm' },
    { "thread-num",     required_argument,  NULL,   'T' },
    { NULL,             0,                  NULL,    0  }
};

static char short_options[] = "hVtdDv:o:c:s:i:a:p:m:T:";

static rstatus_t
nc_daemonize(int dump_core)
{
    rstatus_t status;
    pid_t pid, sid;
    int fd;

    pid = fork();
    switch (pid) {
    case -1:
        log_error("fork() failed: %s", strerror(errno));
        return NC_ERROR;

    case 0:
        break;

    default:
        /* parent terminates */
        _exit(0);
    }

    /* 1st child continues and becomes the session leader */

    sid = setsid();
    if (sid < 0) {
        log_error("setsid() failed: %s", strerror(errno));
        return NC_ERROR;
    }

    if (signal(SIGHUP, SIG_IGN) == SIG_ERR) {
        log_error("signal(SIGHUP, SIG_IGN) failed: %s", strerror(errno));
        return NC_ERROR;
    }

    pid = fork();
    switch (pid) {
    case -1:
        log_error("fork() failed: %s", strerror(errno));
        return NC_ERROR;

    case 0:
        break;

    default:
        /* 1st child terminates */
        _exit(0);
    }

    /* 2nd child continues */

    /* change working directory */
    if (dump_core == 0) {
        status = chdir("/");
        if (status < 0) {
            log_error("chdir(\"/\") failed: %s", strerror(errno));
            return NC_ERROR;
        }
    }

    /* clear file mode creation mask */
    umask(0);

    /* redirect stdin, stdout and stderr to "/dev/null" */

    fd = open("/dev/null", O_RDWR);
    if (fd < 0) {
        log_error("open(\"/dev/null\") failed: %s", strerror(errno));
        return NC_ERROR;
    }

    status = dup2(fd, STDIN_FILENO);
    if (status < 0) {
        log_error("dup2(%d, STDIN) failed: %s", fd, strerror(errno));
        close(fd);
        return NC_ERROR;
    }

    status = dup2(fd, STDOUT_FILENO);
    if (status < 0) {
        log_error("dup2(%d, STDOUT) failed: %s", fd, strerror(errno));
        close(fd);
        return NC_ERROR;
    }

    status = dup2(fd, STDERR_FILENO);
    if (status < 0) {
        log_error("dup2(%d, STDERR) failed: %s", fd, strerror(errno));
        close(fd);
        return NC_ERROR;
    }

    if (fd > STDERR_FILENO) {
        status = close(fd);
        if (status < 0) {
            log_error("close(%d) failed: %s", fd, strerror(errno));
            return NC_ERROR;
        }
    }

    return NC_OK;
}

static void
nc_print_run(struct instance *nci)
{
    int status;
    struct utsname name;

    status = uname(&name);
    if (status < 0) {
        loga("nutcrackers-%s started on pid %d", NC_VERSION_STRING, nci->pid);
    } else {
        loga("nutcracker-%s built for %s %s %s started on pid %d",
             NC_VERSION_STRING, name.sysname, name.release, name.machine,
             nci->pid);
    }

    loga("run, rabbit run / dig that hole, forget the sun / "
         "and when at last the work is done / don't sit down / "
         "it's time to dig another one");
}

static void
nc_print_done(void)
{
    loga("done, rabbit done");
}

static void
nc_show_usage(void)
{
    log_stderr(
        "Usage: nutcrackers [-?hVdDt] [-v verbosity level] [-o output file]" CRLF
        "                   [-c conf file] [-s manage port] [-a manage addr]" CRLF
        "                   [-i interval] [-p pid file] [-m mbuf size]" CRLF
        "                   [-T worker threads number]" CRLF
        "");
    log_stderr(
        "Options:" CRLF
        "  -h, --help             : this help" CRLF
        "  -V, --version          : show version and exit" CRLF
        "  -t, --test-conf        : test configuration for syntax errors and exit" CRLF
        "  -d, --daemonize        : run as a daemon" CRLF
        "  -D, --describe-stats   : print stats description and exit");
    log_stderr(
        "  -v, --verbose=N        : set logging level (default: %d, min: %d, max: %d)" CRLF
        "  -o, --output=S         : set logging file (default: %s)" CRLF
        "  -c, --conf-file=S      : set configuration file (default: %s)" CRLF
        "  -s, --port=N           : set manage port (default: %d)" CRLF
        "  -a, --addr=S           : set manage ip (default: %s)" CRLF
        "  -i, --interval=N       : set interval in msec (default: %d msec)" CRLF
        "  -p, --pid-file=S       : set pid file (default: %s)" CRLF
        "  -m, --mbuf-size=N      : set size of mbuf chunk in bytes (default: %d bytes)" CRLF
        "  -T, --thread_num=N     : set the worker threads number (default: %d)" CRLF
        "",
        NC_LOG_DEFAULT, NC_LOG_MIN, NC_LOG_MAX,
        NC_LOG_PATH != NULL ? NC_LOG_PATH : "stderr",
        NC_CONF_PATH,
        NC_PORT, NC_ADDR, NC_INTERVAL,
        NC_PID_FILE != NULL ? NC_PID_FILE : "off",
        NC_MBUF_SIZE,
        NC_THREAD_NUM_DEFAULT);
}

static rstatus_t
nc_create_pidfile(struct instance *nci)
{
    char pid[NC_UINTMAX_MAXLEN];
    int fd, pid_len;
    ssize_t n;

    fd = open(nci->pid_filename, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd < 0) {
        log_error("opening pid file '%s' failed: %s", nci->pid_filename,
                  strerror(errno));
        return NC_ERROR;
    }
    nci->pidfile = 1;

    pid_len = nc_snprintf(pid, NC_UINTMAX_MAXLEN, "%d", nci->pid);

    n = nc_write(fd, pid, pid_len);
    if (n < 0) {
        log_error("write to pid file '%s' failed: %s", nci->pid_filename,
                  strerror(errno));
        return NC_ERROR;
    }

    close(fd);

    return NC_OK;
}

static void
nc_remove_pidfile(struct instance *nci)
{
    int status;

    status = unlink(nci->pid_filename);
    if (status < 0) {
        log_error("unlink of pid file '%s' failed, ignored: %s",
                  nci->pid_filename, strerror(errno));
    }
}

static void
nc_set_default_options(struct instance *nci)
{
    int status;

    nci->ctx = NULL;

    nci->log_level = NC_LOG_DEFAULT;
    nci->log_filename = NC_LOG_PATH;

    nci->conf_filename = NC_CONF_PATH;

    nci->port = NC_PORT;
    nci->addr = NC_ADDR;
    nci->interval = NC_INTERVAL;

    status = nc_gethostname(nci->hostname, NC_MAXHOSTNAMELEN);
    if (status < 0) {
        log_warn("gethostname failed, ignored: %s", strerror(errno));
        nc_snprintf(nci->hostname, NC_MAXHOSTNAMELEN, "unknown");
    }
    nci->hostname[NC_MAXHOSTNAMELEN - 1] = '\0';

    nci->mbuf_chunk_size = NC_MBUF_SIZE;

    nci->pid = (pid_t)-1;
    nci->pid_filename = NULL;
    nci->pidfile = 0;

    nci->thread_num = (int)NC_THREAD_NUM_DEFAULT;
}

static rstatus_t
nc_get_options(int argc, char **argv, struct instance *nci)
{
    int c, value;

    opterr = 0;

    for (;;) {
        c = getopt_long(argc, argv, short_options, long_options, NULL);
        if (c == -1) {
            /* no more options */
            break;
        }

        switch (c) {
        case 'h':
            show_version = 1;
            show_help = 1;
            break;

        case 'V':
            show_version = 1;
            break;

        case 't':
            test_conf = 1;
            break;

        case 'd':
            daemonize = 1;
            break;

        case 'D':
            describe_stats = 1;
            show_version = 1;
            break;

        case 'v':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcrackers: option -v requires a number");
                return NC_ERROR;
            }
            nci->log_level = value;
            break;

        case 'o':
            nci->log_filename = optarg;
            break;

        case 'c':
            nci->conf_filename = optarg;
            break;

        case 's':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcrackers: option -s requires a number");
                return NC_ERROR;
            }
            if (!nc_valid_port(value)) {
                log_stderr("nutcrackers: option -s value %d is not a valid "
                           "port", value);
                return NC_ERROR;
            }

            nci->port = (uint16_t)value;
            break;

        case 'i':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcrackers: option -i requires a number");
                return NC_ERROR;
            }

            nci->interval = value;
            break;

        case 'a':
            nci->addr = optarg;
            break;

        case 'p':
            nci->pid_filename = optarg;
            break;

        case 'm':
            value = nc_atoi(optarg, strlen(optarg));
            if (value <= 0) {
                log_stderr("nutcrackers: option -m requires a non-zero number");
                return NC_ERROR;
            }

            if (value < NC_MBUF_MIN_SIZE || value > NC_MBUF_MAX_SIZE) {
                log_stderr("nutcrackers: mbuf chunk size must be between %zu and"
                           " %zu bytes", NC_MBUF_MIN_SIZE, NC_MBUF_MAX_SIZE);
                return NC_ERROR;
            }

            nci->mbuf_chunk_size = (size_t)value;
            break;
            
        case 'T':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcrackers: option -T requires a number");
                return NC_ERROR;
            }

            nci->thread_num = value;
            break;

        case '?':
            switch (optopt) {
            case 'o':
            case 'c':
            case 'p':
                log_stderr("nutcrackers: option -%c requires a file name",
                           optopt);
                break;

            case 'm':
            case 'v':
            case 's':
            case 'i':
            case 'T':
                log_stderr("nutcrackers: option -%c requires a number", optopt);
                break;

            case 'a':
                log_stderr("nutcrackers: option -%c requires a string", optopt);
                break;

            default:
                log_stderr("nutcrackers: invalid option -- '%c'", optopt);
                break;
            }
            return NC_ERROR;

        default:
            log_stderr("nutcrackers: invalid option -- '%c'", optopt);
            return NC_ERROR;

        }
    }

    return NC_OK;
}

/*
 * Returns true if configuration file has a valid syntax, otherwise
 * returns false
 */
static bool
nc_test_conf(struct instance *nci, bool test)
{
    struct conf *cf;

    cf = conf_create(nci->conf_filename);
    if (cf == NULL) {
        if (test)
            log_stderr("nutcrackers: configuration file '%s' syntax is invalid",
                nci->conf_filename);
        return false;
    }

    conf_destroy(cf);

    if (test)
        log_stderr("nutcrackers: configuration file '%s' syntax is ok",
            nci->conf_filename);
    return true;
}

static rstatus_t
nc_pre_run(struct instance *nci)
{
    rstatus_t status;

    status = log_init(nci->log_level, nci->log_filename);
    if (status != NC_OK) {
        return status;
    }

    if (!nc_test_conf(nci, false)) {
        log_error("conf file %s is error", nci->conf_filename);
        return NC_ERROR;
    }

    if (daemonize) {
        status = nc_daemonize(1);
        if (status != NC_OK) {
            return status;
        }
    }

    nci->pid = getpid();

    status = signal_init();
    if (status != NC_OK) {
        return status;
    }

    if (nci->pid_filename) {
        status = nc_create_pidfile(nci);
        if (status != NC_OK) {
            return status;
        }
    }

    STATS_LOCK_INIT();

    nc_print_run(nci);

    return NC_OK;
}

static void
nc_post_run(struct instance *nci)
{
    STATS_LOCK_DESTROY();

    if (nci->pidfile) {
        nc_remove_pidfile(nci);
    }

    signal_deinit();

    nc_print_done();

    log_deinit();
}

static void
nc_run(struct instance *nci)
{
    rstatus_t status;
    uint32_t i;
    uint32_t worker_count;
    struct array workers;
    struct thread_data *worker;

    if (nci->thread_num <= 0) {
        log_error("number of work threads must be greater than 0");
        return;
    } else if (nci->thread_num > 64) {
        log_warn("WARNING: Setting a high number of worker threads is not recommended."
            " Set this value to the number of cores in your machine or less.");
    }

    worker_count = (uint32_t)nci->thread_num;
    
    status = array_init(&workers, worker_count, sizeof(struct thread_data));
    if (status != NC_OK) {
        return;
    }
    
    /* init the workers */
    for (i = 0; i < worker_count; i ++) {
        worker = array_push(&workers);
        thread_data_init(worker);
        worker->nci = nci;

        status = setup_worker(worker);
        if (status != NC_OK) {
            exit(1);
        }
    }

    /* run the worker job */
    for (i = 0; i < worker_count; i ++) {
        pthread_attr_t attr;
        pthread_attr_init(&attr);
        worker = array_get(&workers, i);
        pthread_create(&worker->thread_id, 
        	&attr, worker_thread_run, worker);
    }

    status = master_run(nci, &workers);
    if (status != NC_OK) {
        exit(1);
    }

    /* wait for the worker finish */
	for(i = 0; i < worker_count; i ++){
		worker = array_get(&workers, i);
		pthread_join(worker->thread_id, NULL);
	}

    /* deinit the workers */
	for(i = 0; i < worker_count; i ++){
		worker = array_pop(&workers);
		thread_data_deinit(worker);
	}
}

int
main(int argc, char **argv)
{
    rstatus_t status;
    struct instance nci;

    nc_set_default_options(&nci);

    status = nc_get_options(argc, argv, &nci);
    if (status != NC_OK) {
        nc_show_usage();
        exit(1);
    }

    if (show_version) {
        log_stderr("This is nutcrackers-%s" CRLF, NC_VERSION_STRING);
        if (show_help) {
            nc_show_usage();
        }

        if (describe_stats) {
            stats_describe();
        }

        exit(0);
    }

    if (test_conf) {
        if (!nc_test_conf(&nci, true)) {
            exit(1);
        }
        exit(0);
    }

    status = nc_pre_run(&nci);
    if (status != NC_OK) {
        nc_post_run(&nci);
        exit(1);
    }

    nc_run(&nci);

    nc_post_run(&nci);

    exit(1);
}
